linera-sdk/src/lib.rs

```diff
@@ -140,19 +141,6 @@ pub trait Contract: WithContractAbi + ContractAbi + Sized {
     /// For a message to be executed, a user must mark it to be received in a block of the receiver
     /// chain.
     async fn execute_message(&mut self, message: Self::Message) -> Result<(), Self::Error>;
-
-    /// Finishes the execution of the current transaction.
-    ///
-    /// This is called once before a transaction ends, to allow all applications that participated
-    /// in the transaction to perform any final operations, and optionally it may also cancel the
-    /// transaction if there are any pendencies.
-    ///
-    /// The default implementation persists the state, so if this method is overriden, care must be
-    /// taken to persist the state manually.
-    async fn finalize(&mut self) -> Result<(), Self::Error> {
-        Self::State::store(self.state_mut()).await;
```

Comment:
The fact that `finalize` sees `self` and not just `self.state` (the persistent state) is useful for sessions.

Reply:
You can implement `Drop` for either the `Contract` type or the `State` type (or both). So it's still possible to encode sessions, and it can even be coded separately from state validation checks, for example.